tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -0,0 +1,93 @@
+==============================================================================*/
+
+// Provide helper routine for obtaining  gpu target information useful for llvm IR 
+// contruction 
+#include "target_util.h"
+
+#include "llvm/IR/MDBuilder.h"
+
+#include "tensorflow/compiler/xla/service/llvm_ir/llvm_util.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace xla {
+namespace gpu {
+// Obtain target specific llvm intrinsic 
+struct TargetIntrinsics GetIntrinsic(TargetIntrinsicID intrin) {
+  switch (intrin) {
+    case TargetIntrinsicID::kShfldownf32: {
+      return {llvm::Intrinsic::nvvm_shfl_sync_down_f32,
+              llvm::Intrinsic::not_intrinsic};
+    }
+    case TargetIntrinsicID::kShfldowni32: {
+      return {llvm::Intrinsic::nvvm_shfl_sync_down_i32,
+              llvm::Intrinsic::not_intrinsic};
+    }
+    case TargetIntrinsicID::kThreadidx: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_x,
+              llvm::Intrinsic::amdgcn_workitem_id_x};
+    }
+    case TargetIntrinsicID::kThreadidy: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_y,
+              llvm::Intrinsic::amdgcn_workitem_id_y};
+    }
+    case TargetIntrinsicID::kThreadidz: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_z,
+              llvm::Intrinsic::amdgcn_workitem_id_z};
+    }
+    case TargetIntrinsicID::kBlockidx: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_x,
+              llvm::Intrinsic::amdgcn_workgroup_id_x};
+    }
+    case TargetIntrinsicID::kBlockidy: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_y,
+              llvm::Intrinsic::amdgcn_workgroup_id_y};
+    }
+    case TargetIntrinsicID::kBlockidz: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z,
+              llvm::Intrinsic::amdgcn_workgroup_id_z};
+    }
+    case TargetIntrinsicID::kBarrierid: {
+      return {llvm::Intrinsic::nvvm_barrier0,
+              llvm::Intrinsic::amdgcn_s_barrier};
+    }
+  }
+
+  return {llvm::Intrinsic::not_intrinsic, llvm::Intrinsic::not_intrinsic};
+}
+
+llvm::CallInst* EmitCallToTargetIntrinsic(
+    TargetIntrinsicID intrinsic_id, absl::Span<llvm::Value* const> operands,
+    absl::Span<llvm::Type* const> overloaded_types, llvm::IRBuilder<>* b) {
+  llvm::Module* module = b->GetInsertBlock()->getModule();
+  struct TargetIntrinsics gpu_intrinsic_id = GetIntrinsic(intrinsic_id);
+  llvm::Triple target_triple = llvm::Triple(module->getTargetTriple());
+  llvm::Intrinsic::ID llvm_intrinsic_id = llvm::Intrinsic::not_intrinsic;
+  if ((target_triple.getArch() == llvm::Triple::nvptx) ||
+      (target_triple.getArch() == llvm::Triple::nvptx64)) {
+    llvm_intrinsic_id = gpu_intrinsic_id.nvptx_intrinsic;
+  } else if (target_triple.getArch() == llvm::Triple::amdgcn) {
+    llvm_intrinsic_id = gpu_intrinsic_id.amdgpu_intrinsic;
+  } else {
+    LOG(FATAL) << "Invalid triple " << target_triple.str();
+  }
+  CHECK_NE(llvm_intrinsic_id, llvm::Intrinsic::not_intrinsic);
```

Comment:
After changing the function above so it never returns not_intrinsic, I think we can remove this line.